README.md
# Binance Public API Connector DotNET

[![[Nuget]](https://img.shields.io/nuget/v/Binance.Spot)](https://www.nuget.org/packages/Binance.Spot)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

This is a lightweight library that works as a connector to [Binance public API](https://github.com/binance/binance-spot-api-docs)

- Supported APIs:
    - `/api/*`
    - `/sapi/*`
    - Spot WebSocket Market Stream
    - Spot User Data Stream
    - Spot WebSocket API
- Test cases and examples
- Customizable base URL, request timeout and HTTP proxy
- Response Metadata


## Installation

```bash
dotnet add package Binance.Spot
```

## RESTful APIs

Usage example
```csharp
using System;
using System.Threading.Tasks;
using Binance.Spot;

class Program
{
    static async Task Main(string[] args)
    {
        Market market = new Market();

        string serverTime = await market.CheckServerTime();

        Console.WriteLine(serverTime);
    }
}
```

Please find `Examples` folder to check for more endpoints.

## WebSocket Stream

Usage Example
```csharp
using System;
using System.Threading;
using System.Threading.Tasks;
using Binance.Spot;

class Program
{
    static async Task Main(string[] args)
    {
        var websocket = new MarketDataWebSocket("btcusdt@aggTrade");

        websocket.OnMessageReceived(
            (data) =>
        {
            Console.WriteLine(data);

            return Task.CompletedTask;
        }, CancellationToken.None);

        await websocket.ConnectAsync(CancellationToken.None);
    }
}
```
More websocket examples are available in the `Examples` folder

## WebSocket API

Usage Example
```csharp
using System.Threading;
using System.Threading.Tasks;
using Binance.Common;
using Binance.Spot;

public class NewOrder_Example
{
    public static async Task Main(string[] args)
    {
        var websocket = new WebSocketApi("wss://testnet.binance.vision/ws-api/v3", "apiKey", new BinanceHmac("apiSecret"));

        websocket.OnMessageReceived(
            async (data) =>
        {
            Console.WriteLine(data);
            await Task.CompletedTask;
        }, CancellationToken.None);

        await websocket.ConnectAsync(CancellationToken.None);

        await websocket.AccountTrade.NewOrderAsync(symbol: "BNBUSDT", side: Models.Side.BUY, type: Models.OrderType.LIMIT, timeInForce: Models.TimeInForce.GTC, price: 300, quantity: 1, cancellationToken: CancellationToken.None);

        await Task.Delay(5000);
        Console.WriteLine("Disconnect with WebSocket API Server");
        await websocket.DisconnectAsync(CancellationToken.None);
    }
}
```

- The `requestId` param is optional. If not sent, it defaults to a randomly created `GUID` (Globally Unique Identifier).
- The `cancellationToken` is optional. If not sent, it defaults to `CancellationToken.None`.

More websocket API examples are available in the `Examples` folder

## Authentication - RESTful APIs

For API endpoints that requires signature, new authentication interfaces are introduced to generate the signature since V2.

```csharp
// HMAC signature
new SpotAccountTrade(httpClient, new BinanceHmac("apiSecret"), apiKey: "apiKey")

// RSA signature
string privateKey = File.ReadAllText("/Users/john/ssl/PrivateKey.pem");
new SpotAccountTrade(httpClient, new BinanceRsa(privateKey), apiKey: "apiKey")

// Encrypted RSA signature
new SpotAccountTrade(httpClient, new BinanceRsa(privateKey, "thePrivateKeyPassword"), apiKey: "apiKey")

```

For V1.x, it's required to pass `apiKey` and `apiSecret` directly.

```csharp
new SpotAccountTrade(httpClient, apiKey: apiKey, apiSecret: apiSecret)
```

For more details, please find the example from the endpoints `/api/v3/account` in file `AccountInformation_Example`.

## Authentication - WebSocket API

```csharp
// HMAC signature
new WebSocketApi(apiKey: "apiKey", signatureService: new BinanceHmac("apiSecret"));

// RSA signature
string privateKey = File.ReadAllText("/Users/john/ssl/PrivateKey.pem");
new WebSocketApi(apiKey: "apiKey", signatureService: new BinanceRsa(privateKey));

// Encrypted RSA signature
new WebSocketApi(apiKey: "apiKey", signatureService: new BinanceRsa(privateKey, "thePrivateKeyPassword"));

```

### Heartbeat

Once connected, the websocket server sends a ping frame every 3 minutes and requires a response pong frame back within
a 10 minutes period. This package handles the pong responses automatically.

### Testnet

[Spot Testnet](https://testnet.binance.vision/) can be used to test `/api/*` endpoints, Spot Websocket Stream and WebSocket API.

```csharp
using Binance.Spot;

Wallet wallet = new Wallet(baseUrl: "https://testnet.binance.vision");
```

### Base URL

If `baseUrl` is not provided, it defaults to `https://api.binance.com`.

It's recommended to pass in the `baseUrl` parameter, even in production as Binance provides alternative URLs
in case of performance issues:
- `https://api1.binance.com`
- `https://api2.binance.com`
- `https://api3.binance.com`
- `https://api4.binance.com`

For Websocket Stream, `baseUrl` defaults to `wss://stream.binance.com:9443"`.
For Websocket API, `baseUrl` defaults to `wss://ws-api.binance.com:443/ws-api/v3`.

### RecvWindow parameter

Additional parameter `recvWindow` is available for endpoints requiring signature.

It defaults to `5000` (milliseconds) and can be any value lower than `60000`(milliseconds).
Anything beyond the limit will result in an error response from Binance server.

```csharp
using Binance.Spot;

Wallet wallet = new Wallet();

await wallet.AccountStatus(recvWindow=4000)
```

### Timeout
The default timeout is 100,000 milliseconds (100 seconds).  

Usage Example
```csharp
using System;
using System.Net.Http;
using Binance.Spot;

HttpClient httpClient = new HttpClient() { 
    Timeout = TimeSpan.FromSeconds(10)
}

Wallet wallet = new Wallet(httpClient: httpClient);
```

### Proxy
Usage Example
```csharp
using System;
using System.Net;
using System.Net.Http;
using Binance.Spot;

WebProxy proxy = new WebProxy(new Uri("http://1.2.3.4:8080"));
HttpClientHandler proxyHandler = new HttpClientHandler { Proxy = proxy };
HttpClient httpClient = new HttpClient(handler: proxyHandler);

Wallet wallet = new Wallet(httpClient: httpClient);
```

### Exceptions

There are 2 types of exceptions returned from the library:
- `Binance.Common.BinanceClientException`
    - This is thrown when server returns `4XX`, it's an issue from client side.
    - Properties:
        - `Code` - Server's error code, e.g. `-1102`
        - `Message` - Server's error message, e.g. `Unknown order sent.`
- `Binance.Common.BinanceServerException`
    - This is thrown when server returns `5XX`, it's an issue from server side.

Both exceptions inherit `Binance.Common.BinanceHttpException` along with the following properties:
- `StatusCode` - Response http status code, e.g. `401`
- `Headers` -  Dictionary with response headers

### Logging
This library implements the .NET logging API that works with a variety of built-in and third-party logging providers. 

For more information on how to configure logging in .NET, visit [Microsoft's logging article](https://docs.microsoft.com/en-us/dotnet/core/extensions/logging?tabs=command-line)

Usage Example
```csharp
using System;
using System.Net;
using System.Net.Http;
using Binance.Spot;

public async Task LoggingExample(ILogger logger) {
    BinanceLoggingHandler loggingHandler = new BinanceLoggingHandler(logger: logger);
    HttpClient httpClient = new HttpClient(handler: loggingHandler);

    Wallet wallet = new Wallet(httpClient: httpClient);

    await wallet.SystemStatus();
}
```

Sample Output


```
Method: GET, RequestUri: 'https://www.binance.com/?timestamp=1631525776809&signature=f07558c98cb82bcb3556a6a21b8a8a2582bae93d0bb9604a0df72cae8c1c6642', Version: 1.1, Content: <null>, Headers: { }
StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: <null>, Headers: {}
{"status": 0,"msg": "normal"}
```

## Test Cases

```bash
dotnet test
```

## Limitations

Futures and Vanilla Options APIs are not supported:
- /fapi/*
- /dapi/*
- /vapi/*
- Associated WebSocket Market and User Data Streams

## Contributing

Contributions are welcome.

If you've found a bug within this project, please open an issue to discuss what you would like to change.

If it's an issue with the API, please open a topic at [Binance Developer Community](https://dev.binance.vision)


CHANGELOG.md
# Changelog

## 4.0.1 - 2023-10-04

### Updated
- Fix parameter type issue in method `AggregateTradesListAsync`, `KlinesAsync` and `UiKlinesAsync`

## 4.0.0 - 2023-06-28

### Added
- Added User-Agent

## 4.0.0-rc1 - 2023-04-27

### Added
- Added WebSocket API support.
- Added new Models `SelfTradePreventionMode` and `TickerType`.

### Updated
- Updated `MarketDataWebSocket.cs` and `UserDataWebSocket.cs` path to be under `WebSocketStream`.
- Updated `Interval` Model with `ONE_SECOND`.
- Updated dependencies versions:
  - `Microsoft.Extensions.Logging` to `6.0.0`.
  - `Microsoft.Extensions.Logging.Console` to `6.0.0`.
  - `Microsoft.Extensions.Logging.Abstractions` to `6.0.0`.
  - `Microsoft.NET.Test.Sdk` to `17.5.0`.

## 3.0.0 - 2023-03-29

### Removed
- End of compatibility verification for .NET `3.1.x` and `5.0.x` due to these versions reaching their end of life.

### Updated
- Replaced `BouncyCastle` cryptographic package to `BouncyCastle.Cryptography`.
- Updated Examples and Tests project file to `net6.0`.

## 2.0.0 - 2022-11-22

### Added
- Add RSA Key signature

## 1.8.0 - 2022-10-11

### Added
- New endpoint for Portfolio Margin:
  - `GET /sapi/v1/portfolio/pmLoan` to query Portfolio Margin Bankruptcy Loan Record.
  - `POST /sapi/v1/portfolio/repay` to repay Portfolio Margin Bankruptcy Loan.
  - `GET /sapi/v1/portfolio/collateralRate` to get Portfolio Margin Collateral Rate.

### Changed
- Changed endpoints for Trade:
  - `POST /api/v3/order/test` New optional fields `strategyId` and `strategyType`.
  - `POST /api/v3/order` New optional fields `strategyId` and `strategyType`.
  - `POST /api/v3/order/cancelReplace` New optional fields `strategyId` and `strategyType`.
  - `POST /api/v3/order/oco` New optional fields `limitStrategyId`, `limitStrategyType`, `stopStrategyId` and `stopStrategyType`.
- Fixing null value appended in the stream
- Update WebSocket example code

## 1.7.0 - 2022-07-19

### Added
- New endpoint for Margin:
  - `POST /sapi/v3/asset/getUserAsset` to get user assets.

- New endpoint for Wallet:
  - `GET /sapi/v1/margin/dribblet` to query the historical information of user's margin account small-value asset conversion BNB.

### Changed
- Update endpoint for Convert:
  - `GET /sapi/v1/fiat/orders`: Weight changes from IP(3000) to IP(100)

## 1.6.0 - 2022-07-04

### Added
- New endpoint for Market:
  - `GET /api/v3/ticker` for rolling window price change statistics based on windowSize provided.

- New endpoint for Account Trade:
  - `POST /api/v3/order/cancelReplace` to cancel an existing order and place a new order on the same symbol.

### Changed
- Update endpoint for Fiat:
  - `GET /sapi/v1/fiat/orders`: Weight changes from IP(1) to UID(90000)

- Update endpoint for Pay:
  - `GET /sapi/v1/pay/transactions`: Param names changed: startTimestamp -> startTime; endTimestamp -> endTime.

### Fixed
- Included `Examples/CSharp` and `Examples/FSharp` build process in CI pipeline.

- Explicitly set optional parameters in examples.

## 1.5.0 - 2022-05-27
### Added
- New endpoint for Binance Code:
  - `GET /sapi/v1/giftcard/cryptography/rsa-public-key` to fetch RSA public key.

- New endpoints for Staking:
  - `GET /sapi/v1/staking/productList` to get Staking product list
  - `POST /sapi/v1/staking/purchase` to stake product
  -   - `POST /sapi/v1/staking/redeem` to redeem product
  - `GET /sapi/v1/staking/position` to get Staking product holding position
  - `GET /sapi/v1/staking/stakingRecord` to inquiry Staking history records
  - `POST /sapi/v1/staking/setAutoStaking` to set Auto Staking function
  - `GET /sapi/v1/staking/personalLeftQuota` to inquiry Staking left quota

- New endpoint for Portfolio Margin:
  - `GET /sapi/v1/portfolio/account` to support query portfolio margin account info

- New endpoint for BSwap:
  - `GET /sapi/v1/bswap/removeLiquidityPreview` to get remove liquidity preview

### Changed
- Update endpoint for Binance Code:
  - `POST /sapi/v1/giftcard/redeemCode`: new optional parameter externalUid. Each external unique ID represents a unique user on the partner platform. The function helps you to identify the redemption behavior of different users.

- Update endpoints for Market:
  - `GET /api/v3/ticker/24hr`, `GET /api/v3/ticker/price` and `GET /api/v3/ticker/bookTicker` new optional parameter symbols.

### Fixed
- Added missing `Market` examples.
- Updated endpoint documentation and examples to latest version.
- Added missing `Mining` endpoint parameters.

## 1.4.0 - 2022-04-20
### Added
- New endpoint for Sub-Account
  - `GET /sapi/v1/managed-subaccount/accountSnapshot` to support investor master account query asset snapshot of managed sub-account.
- New transfer types `MARGIN`, `ISOLATED_MARGIN` and parameter `symbol` for `POST /sapi/v1/sub-account/universalTransfer` to support transfer to sub-account cross margin account and isolated margin account.

### Changed
- Changed endpoints for Spot Trading
  - `POST /api/v3/order` - New optional field trailingDelta.
  - `POST /api/v3/order/test` - New optional field trailingDelta.
  - `POST /api/v3/order/oco` - New optional field trailingDelta.
- Changed endpoints for Sub-Account
- `GET /sapi/v1/sub-account/universalTransfer`
  - The query time period must be less then 30 days; If startTime and endTime not sent, return records of the last 30 days by default.
- `GET api/v3/depth`
  - The limit value can be outside of the previous values (i.e. 5, 10, 20, 50, 100, 500, 1000,5000) and will return the correct limit. (i.e. if limit=3 then the response will be the top 3 bids and asks)
  - The limit still cannot exceed 5000. If the limit provided is greater than 5000, then the response will be truncated to 5000.

### Fixed
- Fixed query parameters for `POST /api/v3/order/test`.

## 1.3.1 - 2022-03-08
### Fixed
- Query parameter is no longer subject to culture.
- Trade examples have more appropriate arguments

## 1.3.0 - 2022-03-02
### Added
- New endpoints for Wallet

Src/Spot/Market.cs
namespace Binance.Spot
{
    using System.Collections.Generic;
    using System.Net.Http;
    using System.Threading.Tasks;
    using Binance.Common;
    using Binance.Spot.Models;

    public class Market : SpotService
    {
        public Market(string baseUrl = DEFAULT_SPOT_BASE_URL, string apiKey = null, string apiSecret = null)
        : this(new HttpClient(), baseUrl: baseUrl, apiKey: apiKey, apiSecret: apiSecret)
        {
        }

        public Market(HttpClient httpClient, string baseUrl = DEFAULT_SPOT_BASE_URL, string apiKey = null, string apiSecret = null)
        : base(httpClient, baseUrl: baseUrl, apiKey: apiKey, apiSecret: apiSecret)
        {
        }

        public Market(HttpClient httpClient, IBinanceSignatureService signatureService, string baseUrl = DEFAULT_SPOT_BASE_URL, string apiKey = null)
        : base(httpClient, baseUrl: baseUrl, apiKey: apiKey, signatureService: signatureService)
        {
        }

        private const string TEST_CONNECTIVITY = "/api/v3/ping";

        /// <summary>
        /// Test connectivity to the Rest API.<para />
        /// Weight(IP): 1.
        /// </summary>
        /// <returns>OK.</returns>
        public async Task<string> TestConnectivity()
        {
            var result = await this.SendPublicAsync<string>(
                TEST_CONNECTIVITY,
                HttpMethod.Get);

            return result;
        }

        private const string CHECK_SERVER_TIME = "/api/v3/time";

        /// <summary>
        /// Test connectivity to the Rest API and get the current server time.<para />
        /// Weight(IP): 1.
        /// </summary>
        /// <returns>Binance server UTC timestamp.</returns>
        public async Task<string> CheckServerTime()
        {
            var result = await this.SendPublicAsync<string>(
                CHECK_SERVER_TIME,
                HttpMethod.Get);

            return result;
        }

        private const string EXCHANGE_INFORMATION = "/api/v3/exchangeInfo";

        /// <summary>
        /// Current exchange trading rules and symbol information.<para />
        /// - If any symbol provided in either symbol or symbols do not exist, the endpoint will throw an error.<para />
        /// Weight(IP): 10.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="symbols"></param>
        /// <returns>Current exchange trading rules and symbol information.</returns>
        public async Task<string> ExchangeInformation(string symbol = null, string symbols = null)
        {
            var result = await this.SendPublicAsync<string>(
                EXCHANGE_INFORMATION,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "symbols", symbols },
                });

            return result;
        }

        private const string ORDER_BOOK = "/api/v3/depth";

        /// <summary>
        /// | Limit               | Weight(IP)  |.<para />
        /// |---------------------|-------------|.<para />
        /// | 1-100               | 1           |.<para />
        /// | 101-500             | 5           |.<para />
        /// | 501-1000            | 10          |.<para />
        /// | 1001-5000           | 50          |.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="limit">If limit > 5000, then the response will truncate to 5000.</param>
        /// <returns>Order book.</returns>
        public async Task<string> OrderBook(string symbol, int? limit = null)
        {
            var result = await this.SendPublicAsync<string>(
                ORDER_BOOK,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "limit", limit },
                });

            return result;
        }

        private const string RECENT_TRADES_LIST = "/api/v3/trades";

        /// <summary>
        /// Get recent trades.<para />
        /// Weight(IP): 1.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="limit">Default 500; max 1000.</param>
        /// <returns>Trade list.</returns>
        public async Task<string> RecentTradesList(string symbol, int? limit = null)
        {
            var result = await this.SendPublicAsync<string>(
                RECENT_TRADES_LIST,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "limit", limit },
                });

            return result;
        }

        private const string OLD_TRADE_LOOKUP = "/api/v3/historicalTrades";

        /// <summary>
        /// Get older market trades.<para />
        /// Weight(IP): 5.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="limit">Default 500; max 1000.</param>
        /// <param name="fromId">Trade id to fetch from. Default gets most recent trades.</param>
        /// <returns>Trade list.</returns>
        public async Task<string> OldTradeLookup(string symbol, int? limit = null, long? fromId = null)
        {
            var result = await this.SendPublicAsync<string>(
                OLD_TRADE_LOOKUP,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "limit", limit },
                    { "fromId", fromId },
                });

            return result;
        }

        private const string COMPRESSED_AGGREGATE_TRADES_LIST = "/api/v3/aggTrades";

        /// <summary>
        /// Get compressed, aggregate trades. Trades that fill at the time, from the same order, with the same price will have the quantity aggregated.<para />
        /// - If `startTime` and `endTime` are sent, time between startTime and endTime must be less than 1 hour.<para />
        /// - If `fromId`, `startTime`, and `endTime` are not sent, the most recent aggregate trades will be returned.<para />
        /// Weight(IP): 1.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="fromId">Trade id to fetch from. Default gets most recent trades.</param>
        /// <param name="startTime">UTC timestamp in ms.</param>
        /// <param name="endTime">UTC timestamp in ms.</param>
        /// <param name="limit">Default 500; max 1000.</param>
        /// <returns>Trade list.</returns>
        public async Task<string> CompressedAggregateTradesList(string symbol, long? fromId = null, long? startTime = null, long? endTime = null, int? limit = null)
        {
            var result = await this.SendPublicAsync<string>(
                COMPRESSED_AGGREGATE_TRADES_LIST,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "fromId", fromId },
                    { "startTime", startTime },
                    { "endTime", endTime },
                    { "limit", limit },
                });

            return result;
        }

        private const string KLINE_CANDLESTICK_DATA = "/api/v3/klines";

        /// <summary>
        /// Kline/candlestick bars for a symbol.<para />
        /// Klines are uniquely identified by their open time.<para />
        /// - If `startTime` and `endTime` are not sent, the most recent klines are returned.<para />
        /// Weight(IP): 1.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="interval">kline intervals.</param>
        /// <param name="startTime">UTC timestamp in ms.</param>
        /// <param name="endTime">UTC timestamp in ms.</param>
        /// <param name="limit">Default 500; max 1000.</param>
        /// <returns>Kline data.</returns>
        public async Task<string> KlineCandlestickData(string symbol, Interval interval, long? startTime = null, long? endTime = null, int? limit = null)
        {
            var result = await this.SendPublicAsync<string>(
                KLINE_CANDLESTICK_DATA,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "interval", interval },
                    { "startTime", startTime },
                    { "endTime", endTime },
                    { "limit", limit },
                });

            return result;
        }

        private const string CURRENT_AVERAGE_PRICE = "/api/v3/avgPrice";

        /// <summary>
        /// Current average price for a symbol.<para />
        /// Weight(IP): 1.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <returns>Average price.</returns>
        public async Task<string> CurrentAveragePrice(string symbol)
        {
            var result = await this.SendPublicAsync<string>(
                CURRENT_AVERAGE_PRICE,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                });

            return result;
        }

        private const string TWENTY_FOUR_HR_TICKER_PRICE_CHANGE_STATISTICS = "/api/v3/ticker/24hr";

        /// <summary>
        /// 24 hour rolling window price change statistics. Careful when accessing this with no symbol.<para />
        /// - If the symbol is not sent, tickers for all symbols will be returned in an array.<para />
        /// Weight(IP):.<para />
        /// - `1` for a single symbol;.<para />
        /// - `40` when the symbol parameter is omitted;.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="symbols"></param>
        /// <returns>24hr ticker.</returns>
        public async Task<string> TwentyFourHrTickerPriceChangeStatistics(string symbol = null, string symbols = null)
        {
            var result = await this.SendPublicAsync<string>(
                TWENTY_FOUR_HR_TICKER_PRICE_CHANGE_STATISTICS,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "symbols", symbols },
                });

            return result;
        }

        private const string SYMBOL_PRICE_TICKER = "/api/v3/ticker/price";

        /// <summary>
        /// Latest price for a symbol or symbols.<para />
        /// - If the symbol is not sent, prices for all symbols will be returned in an array.<para />
        /// Weight(IP):.<para />
        /// - `1` for a single symbol;.<para />
        /// - `2` when the symbol parameter is omitted;.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="symbols"></param>
        /// <returns>Price ticker.</returns>
        public async Task<string> SymbolPriceTicker(string symbol = null, string symbols = null)
        {
            var result = await this.SendPublicAsync<string>(
                SYMBOL_PRICE_TICKER,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "symbols", symbols },
                });

            return result;
        }

        private const string SYMBOL_ORDER_BOOK_TICKER = "/api/v3/ticker/bookTicker";

        /// <summary>
        /// Best price/qty on the order book for a symbol or symbols.<para />
        /// - If the symbol is not sent, bookTickers for all symbols will be returned in an array.<para />
        /// Weight(IP):.<para />
        /// - `1` for a single symbol;.<para />
        /// - `2` when the symbol parameter is omitted;.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="symbols"></param>
        /// <returns>Order book ticker.</returns>
        public async Task<string> SymbolOrderBookTicker(string symbol = null, string symbols = null)
        {
            var result = await this.SendPublicAsync<string>(
                SYMBOL_ORDER_BOOK_TICKER,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "symbols", symbols },
                });

            return result;
        }

        private const string ROLLING_WINDOW_PRICE_CHANGE_STATISTICS = "/api/v3/ticker";

        /// <summary>
        /// The window used to compute statistics is typically slightly wider than requested windowSize.<para />
        /// openTime for /api/v3/ticker always starts on a minute, while the closeTime is the current time of the request. As such, the effective window might be up to 1 minute wider than requested.<para />
        /// E.g. If the closeTime is 1641287867099 (January 04, 2022 09:17:47:099 UTC) , and the windowSize is 1d. the openTime will be: 1641201420000 (January 3, 2022, 09:17:00 UTC).<para />
        /// Weight(IP): 2 for each requested symbol regardless of windowSize.<para />
        /// The weight for this request will cap at 100 once the number of symbols in the request is more than 50.
        /// </summary>
        /// <param name="symbol">Trading symbol, e.g. BNBUSDT.</param>
        /// <param name="symbols">Either symbol or symbols must be provided.<para />
        /// Examples of accepted format for the symbols parameter: ["BTCUSDT","BNBUSDT"] or %5B%22BTCUSDT%22,%22BNBUSDT%22%5D.<para />
        /// The maximum number of symbols allowed in a request is 100.</param>
        /// <param name="windowSize">Defaults to 1d if no parameter provided.<para />
        /// Supported windowSize values:.<para />
        /// 1m,2m....59m for minutes.<para />
        /// 1h, 2h....23h - for hours.<para />
        /// 1d...7d - for days.<para />
        /// Units cannot be combined (e.g. 1d2h is not allowed).</param>
        /// <returns>Rolling price ticker.</returns>
        public async Task<string> RollingWindowPriceChangeStatistics(string symbol = null, string symbols = null, string windowSize = null)
        {
            var result = await this.SendPublicAsync<string>(
                ROLLING_WINDOW_PRICE_CHANGE_STATISTICS,
                HttpMethod.Get,
                query: new Dictionary<string, object>
                {
                    { "symbol", symbol },
                    { "symbols", symbols },
                    { "windowSize", windowSize },
                });

            return result;
        }
    }
}